File: src/cpp/lib/testml/run/tap.cpp
#include <string>
#include <iostream>
#include "tap.hpp"

namespace testml {
namespace run {

  void TAP::testml_eq(json want, json got, std::string const& label) {
    if (want == got) {
      tap_pass(label);
    } else {
      tap_fail(label);
    }
  }

  void TAP::tap_pass(std::string const& label) {
    std::cout << "ok " << ++count;
    if (!label.empty()) {
      std::cout << " - " << label;
    }
    std::cout << "\n";
  }

  void TAP::tap_fail(std::string const& label) {
    std::cout << "not ok " << ++count;
    if (!label.empty()) {
      std::cout << " - " << label;
    }
    std::cout << "\n";
  }

  void TAP::testml_end() {
    std::cout << "1.." << count;
  }

}
}

File: src/elixir/lib/testml/run/tap.ex
defmodule TestML.Run.Tap do
  alias __MODULE__
  alias TestML.Run

  defstruct \
    count: 0,
    checked: false,
    planned: false

  def run(file) do
    Tap.new(file)
      |> Run.from_file(file)
      |> Run.test
  end

  def new(file) do
    %Tap{} |> Run.new(file)
  end

  def testml_begin(%Tap{} = self) do
    self = %{self | checked: false}
    self = %{self | planned: false}
    self
  end

  def testml_end(%Tap{} = self) do
    if not self.planned do
      self|>tap_done
    end
    self
  end

  def testml_eq_fake(%Tap{} = self) do
    self = Map.put(self, :count, self.count + 1)
    # self = %{self | count: self.count + 1}
    IO.puts("ok #{self.count}")
    {:reply, self}
  end

  def tap_done(%Tap{} = self) do
    # count = to_string(self.count)
    count = "1"  # XXX
    IO.puts("1..#{count}")

    self
  end

end
